File: 0155-min-stack/0155-min-stack.cpp
class MinStack
{
private:
    stack<int> stk;
    stack<int> minstk;

public:
    MinStack()
    {
    }

    void push(int val)
    {
        stk.push(val);
        if (!minstk.empty() && minstk.top() <= val)
        {
            int minVal = minstk.top();
            minstk.push(minVal);
        }
        else
        {
            minstk.push(val);
        }
    }

    void pop()
    {
        if (!stk.empty())
        {
            minstk.pop();
            stk.pop();
        }
    }

    int top()
    {
        if (stk.top())
            return stk.top();
        return 0;
    }

    int getMin()
    {
        if (minstk.empty())
            return 0;
        int minVal = minstk.top();
        return minVal;
    }
};

File: 0155-min-stack/0155-min-stack.java
import java.util.AbstractMap;
import java.util.Map;
import java.util.Stack;

class MinStack {
    Stack<Map.Entry<Integer, Integer>> stack;
    // Entry key is the inserted value into the stack, Entry value is the min value in the stack at this time


    public MinStack() {
        stack = new Stack<>();
    }

    public void push(int val) {
        int minValue = Math.min(stack.empty() ? val : stack.peek().getValue(), val);
        stack.push(new AbstractMap.SimpleImmutableEntry<>(val,minValue));
    }

    public void pop() {
        stack.pop();
    }

    public int top() {
        return stack.peek().getKey();
    }

    public int getMin() {
        return stack.peek().getValue();
    }
}